CU-868az54kh: Initial SceneExporter, Window and CompressionUtility

adrenak opened the pull request.

nic-mxr commented on Assets/MXR.SDK/Runtime/Utils/ZipUtils.cs:
```diff
@@ -0,0 +1,85 @@
+
+using UnityEngine;
+
+public static class ZipUtils {
```
Is this coming from the Zip Lib or something we made?

It would be nice if we used some type of `IZipUtility` interface with public methods for `Compress` and `Extract`. This would mean if we wanted to use a different library in the future there would be less directly coupling with this 3rd part code.

adrenak replied:
Good callout! Put it behind `ICompressionUtility`. I've removed the usage of ZipUtils but the class is still there for now

adrenak commented on Assets/MXR.SDK/Editor/Scene Export/SceneExporter.cs:
```diff
@@ -0,0 +1,71 @@
+namespace MXR.SDK.Editor {
+    public static class SceneExporter {
+        /// <summary>
+        /// Exports a scene into a zip arachive containing three asset bundles and their manifests:
+        /// - "assets" containing the asset dependencies such as textures, models, materials, shaders
+        /// - "scene" containing the scene
+        /// - a third named after the output path with general metadata
+        /// </summary>
```
This is a metadata bundle that Unity generates along with scene and assets automatically. Its name cannot be defined and is named after the directory the other two bundles are exported to.

adrenak commented on Assets/MXR.SDK/Editor/Scene Export/SceneExporter.cs:
```diff
@@ -0,0 +1,234 @@
+            // Create files.txt containing the dependencies inside the export directory
+            File.WriteAllText(Path.Combine(exportDir, "build_report.txt"), buildReport.ToPrettyString());
+
+            // Delete the export directory if required
+            if (deleteExportDir)
+                Directory.Delete(exportDir, recursive: true);
+
+            // Compress export directory to the output file path
+            ICompressionUtility zipCompression = new SharpZipLibCompressionUtility();
+            zipCompression.CompressDirectory(exportDir, outputFilePath);
```
Good bot.

adrenak commented on Assets/MXR.SDK/Editor/Extensions.cs:
```diff
@@ -0,0 +1,109 @@
+        public static string ToPrettyString(this BuildReport buildReport, int indentLength = 4) {
+            StringBuilder sb = new StringBuilder();
+            int indentLevel = 0;
+            void Append(string text, bool newLine = true) {
+                sb.Append(' ', indentLevel * indentLength);
+                sb.Append(text);
+                if (newLine)
+                    sb.Append("\n");
+            }
```
Kinda disagree here. The Append method and the indentLevel variable are key to understanding how the rest of the method indents the lines.

adrenak commented:
@nic-mxr your comment about adding an interface for ZipUtils got lost when I moved it. I've introduced ICompressionUtility interface. ZipUtils.cs is still 

Also I've moved some code out of SceneExporter to Utils and Extensions for clarity. Extensions.cs has a function to pretty print the BuildReport cleanly that also gets packaged with the export now. Serializing BuildReport with JsonUtility didn't work due to some incompatible Unity types inside it, neither did a reflection based method that would recursively traverse the nested objects because there are public properties in its hierarchy that are obsolete.

adrenak commented on Assets/MXR.SDK/Editor/Scene Export/SceneExporter.cs:
```diff
@@ -0,0 +1,124 @@
+            
+            // If build failed, write failed report and show an error dialog
+            if (!DidBuildSucceed(manifest, buildReport)) {
+                if(buildReport != null) {
```
Very minor. Will get cleaned up eventually in later commits anyway.

nic-mxr commented on Assets/MXR.SDK/Runtime/Utils/Compression/ZipUtils.cs:
```diff
@@ -0,0 +1,19 @@
+﻿namespace MXR.SDK {
+    public static class ZipUtils {
+        readonly static ICompressionUtility _compressionUtility = new SharpZipLibCompressionUtility();
```
This is definitely a valid approach, it might also be nice to have some type of "Set" method to provide a new implementation, even if we default to this one. That way if we write tests for things that use this, then we could provide a mock and still be testable.

adrenak replied:
Hmm the branches upstack create the ICompressionUtility object where needed and use that directly. We could get rid of this class altogether unless you think keeping it adds some value.

adrenak replied:
Seeing less and less reason to keep this file as we can just create the required ICompressionUtility object instead. We can remove this altogether in a polish pass later

nic-mxr commented:
Took another look!

Excited to see the interface definition, hope that helps!

Left a comment that might help to further improve testability. Other than that I feel like the main thing is just hitting codestyle, mostly just need to add `private` access modifier to a lot of things - didn't want tag them all as that would be overwhelming.

adrenak commented:
I have made the codestyle changes up the stack. Probably should have started with them now that I think of it.
